compiler/lib/global_deadcode.ml

```diff
@@ -0,0 +1,445 @@
+(* Js_of_ocaml compiler
+ * http://www.ocsigen.org/js_of_ocaml/
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, with linking exception;
+ * either version 2.1 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ *)
+
+open Code
+open Stdlib
+
+let debug = Debug.find "globaldeadcode"
+
+let times = Debug.find "times"
+
+(** Definition of a variable [x]. *)
+type def =
+  | Expr of expr  (** [x] is defined by an expression. *)
+  | Param  (** [x] is a block or closure parameter. *)
+
+(** Liveness of a variable [x], forming a lattice structure. *)
+type live =
+  | Top  (** [x] is live and not a block. *)
+  | Live of IntSet.t  (** [x] is a live block with a (non-empty) set of live fields. *)
+  | Dead  (** [x] is dead. *)
+
+module G = Dgraph.Make_Imperative (Var) (Var.ISet) (Var.Tbl)
+
+module Domain = struct
+  type t = live
+
+  let equal l1 l2 =
+    match l1, l2 with
+    | Top, Top | Dead, Dead -> true
+    | Live f1, Live f2 -> IntSet.equal f1 f2
+    | Top, (Dead | Live _) | Live _, (Dead | Top) | Dead, (Live _ | Top) -> false
+
+  let bot = Dead
+
+  (** Join the liveness according to lattice structure. *)
+  let join l1 l2 =
+    match l1, l2 with
+    | _, Top | Top, _ -> Top
+    | Live f1, Live f2 -> Live (IntSet.union f1 f2)
+    | Dead, Live f | Live f, Dead -> Live f
+    | Dead, Dead -> Dead
+end
+
+module Solver = G.Solver (Domain)
+
+let pure_expr pure_funs e = Pure_fun.pure_expr pure_funs e
```

Comment:
This alias is not necessary

Reply:
[48fccd4](https://github.com/ocsigen/js_of_ocaml/pull/1503/commits/48fccd4502c0114a26b0c3595dc37773a74a6c42)